n2v/methods/zmp.py

```diff
@@ -33,174 +35,269 @@ def zmp(self,
-
-            #Equation 7 of Reference (1)
-            v_c = lam * ( (J[0] + J[1]) - (self.J0[0] + self.J0[1]) )
-
-            #Equation 10 of Reference (1)
-            Fa += H + self.va + v_c 
-            Fb += H + self.vb + v_c
-
-            #Level Shift
-            Fa += (S2 - reduce(np.dot, (S2, Da, S2)) * self.shift)
-            Fb += (S2 - reduce(np.dot, (S2, Db, S2)) * self.shift)
-
-#------------->  DIIS:
-            if SCF_ITER > 1:
-                #Construct the AO gradient
-                # r = (A(FDS - SDF)A)_mu_nu
-                grad_a = A.dot(Fa.dot(Da).dot(S2) - S2.dot(Da).dot(Fa)).dot(A)
-                grad_a[np.abs(grad_a) < eps] = 0.0
-
-                if SCF_ITER -1 < self.diis_space:
-                    state_vectors_a.append(Fa.copy())
-                    error_vectors_a.append(grad_a.copy())
-                else:
-                    state_vectors_a.append(Fa.copy())
-                    error_vectors_a.append(grad_a.copy())
-
-                #Build inner product of error vectors
-                Bdim = len(state_vectors_a)
-                Ba = np.empty((Bdim + 1, Bdim + 1))
-                Ba[-1, :] = -1
-                Ba[:, -1] = -1
-                Ba[-1, -1] = 0
-                Bb = Ba.copy()
-
-                for i in range(len(state_vectors_a)):
-                    for j in range(len(state_vectors_a)):
-                        Ba[i,j] = np.einsum('ij,ij->', error_vectors_a[i], error_vectors_a[j])
-
-                #Build almost zeros matrix to generate inverse. 
-                RHS = np.zeros(Bdim + 1)
-                RHS[-1] = -1
-
-                #Find coefficient matrix:
-                Ca = np.linalg.solve(Ba, RHS.copy())
-                Ca[np.abs(Ca) < eps] = 0.0 
-
-                #Generate new fock Matrix:
-                Fa = np.zeros_like(Fa)
-                for i in range(Ca.shape[0] - 1):
-                    Fa += Ca[i] * state_vectors_a[i]
-
-                diis_error_a = np.max(np.abs(error_vectors_a[-1]))
-                if self.ref ==  1:
-                    Fb = Fa.copy()
-                    diis_error = 2 * diis_error_a
+                Fa = np.zeros((self.nbf, self.nbf))
+                Fb = np.zeros((self.nbf, self.nbf))
+                J, _ = self.form_jk(Cocca, Coccb)
 
+                if zmp_functional.lower() == 'hartree':
+                    #Equation 7 of Reference (1)
+                    vc = lam_i * ( (J[0] + J[1]) - (self.J0[0] + self.J0[1]) )
                 else:
-                    grad_b = A.dot(Fb.dot(Db).dot(S2) - S2.dot(Db).dot(Fb)).dot(A)
-                    grad_b[np.abs(grad_b) < eps] = 0.0
-                    
-                    if SCF_ITER -1 < self.diis_space:
-                        state_vectors_b.append(Fb.copy())
-                        error_vectors_b.append(grad_b.copy())
+                    #Equations 37, 38, 39, 40 of Reference (3)
+                    vc = self.generate_s_fucntional(lam_i, zmp_functional)
```

Comment:
Would it be better to also include condition zmp_functional.lower() == 'hartree' in generate_s_fucntional() instead of picking this out?

Reply:
Makes sense. Will put inside